src/tpm/tpm_wrapper.rs

```diff
@@ -0,0 +1,310 @@
+use tpm::Error as TpmError;
+
+use tss2::{
+    Esys_ContextLoad, Esys_ECDH_ZGen, Esys_Finalize, Esys_FlushContext, Esys_Initialize,
+    Esys_ReadPublic, Fapi_Finalize, Fapi_GetEsysBlob, Fapi_GetTcti, Fapi_Initialize, Fapi_Sign,
+    Tss2_MU_TPMS_CONTEXT_Unmarshal, ESYS_CONTEXT, ESYS_TR, ESYS_TR_NONE, ESYS_TR_PASSWORD,
+    FAPI_CONTEXT, FAPI_ESYSBLOB_CONTEXTLOAD, TPM2B_ECC_POINT, TPM2B_NAME, TPM2B_PUBLIC,
+    TPMS_CONTEXT, TSS2_RC, TSS2_RC_SUCCESS, TSS2_TCTI_CONTEXT, UINT16,
+};
+
+pub type Result<T = ()> = std::result::Result<T, error::Error>;
+
+static mut TPM_CTX: Option<Mutex<*mut FAPI_CONTEXT>> = None;
+
+fn tpm<'a>() -> &'a Mutex<*mut FAPI_CONTEXT> {
+    unsafe { TPM_CTX.as_ref().unwrap() }
+}
+
+fn with_tpm<F, R>(f: F) -> R
+where
+    F: FnOnce(*mut FAPI_CONTEXT) -> R,
+{
+    let tpm_ctx = tpm().lock().unwrap();
+    f(*tpm_ctx)
+}
+
+pub fn tpm_init() -> Result {
+    unsafe {
+        let mut tpm_ctx: *mut FAPI_CONTEXT = null_mut();
+        let res = Fapi_Initialize(&mut tpm_ctx as *mut *mut FAPI_CONTEXT, null_mut());
+        if res != TSS2_RC_SUCCESS {
+            return Err(TpmError::tpm_error("Fapi_Initialize", res));
+        }
+
+        TPM_CTX = Some(Mutex::new(tpm_ctx));
+    }
+
+    Ok(())
+}
+
+pub fn tpm_deinit() {
+    unsafe {
+        let mut tpm_ctx = *tpm().lock().unwrap();
+        Fapi_Finalize(&mut tpm_ctx);
+        TPM_CTX = None;
+    }
+}
+
+fn free_esys_resources(esys_ctx: &mut *mut ESYS_CONTEXT, esys_key_handle: ESYS_TR) -> Result {
+    unsafe {
+        if esys_key_handle != u32::MAX {
+            let res = Esys_FlushContext(*esys_ctx, esys_key_handle);
+            if res != TSS2_RC_SUCCESS {
+                return Err(TpmError::tpm_error("Esys_FlushContext", res));
+            }
+        }
+
+        Esys_Finalize(&mut *esys_ctx);
+    }
+
+    Ok(())
+}
+
+pub fn public_key(key_path: &str) -> Result<Vec<u8>> {
+    unsafe {
+        let tpm_ctx = tpm().lock().unwrap();
+        let mut esys_key_handle: ESYS_TR = u32::MAX;
+        let mut blob_type: u8 = 0;
+        let mut esys_blob: *mut u8 = null_mut();
+        let mut blob_sz: tss2::size_t = 0;
+        let mut offset: tss2::size_t = 0;
+        let mut result = Fapi_GetEsysBlob(
+            *tpm_ctx,
+            CString::new(key_path.as_bytes()).unwrap().as_ptr(),
+            &mut blob_type as *mut u8,
+            &mut esys_blob as *mut *mut u8,
+            &mut blob_sz as *mut tss2::size_t,
+        );
+        if result != TSS2_RC_SUCCESS {
+            return Err(TpmError::tpm_error("Fapi_GetEsysBlob", result));
+        }
+        if blob_type != FAPI_ESYSBLOB_CONTEXTLOAD as u8 {
+            return Err(TpmError::wrong_key_path());
+        }
+
+        let mut key_context: TPMS_CONTEXT = MaybeUninit::uninit().assume_init();
+        result = Tss2_MU_TPMS_CONTEXT_Unmarshal(
+            esys_blob,
+            blob_sz,
+            &mut offset as *mut tss2::size_t,
+            &mut key_context as *mut TPMS_CONTEXT,
+        );
```

Comment:
I haven't tried this, and would need to read up MaybeUnit again, but I suspect the correct usage of MaybeUnit here is like so:
```suggestion
        let mut key_context: TPMS_CONTEXT = MaybeUninit::uninit();
        result = Tss2_MU_TPMS_CONTEXT_Unmarshal(
            esys_blob,
            blob_sz,
            &mut offset as *mut tss2::size_t,
            key_context.as_mut_ptr(),
        );
        key_context.assume_init();
```